source/common/router/vhds.h

```diff
@@ -0,0 +1,100 @@
+        const envoy::api::v2::core::ConfigSource&, const LocalInfo::LocalInfo&, Event::Dispatcher&,
+        Upstream::ClusterManager&, Envoy::Runtime::RandomGenerator&, Stats::Scope&, const std::string&,
+        const std::string&, absl::string_view, Api::Api&);
+
+class VhdsSubscription : Envoy::Config::SubscriptionCallbacks,
+                         Logger::Loggable<Logger::Id::router>,
+                         public RouteConfigUpdateInfo {
+public:
+  VhdsSubscription(const envoy::api::v2::RouteConfiguration& route_configuration,
+                   Server::Configuration::FactoryContext& factory_context,
+                   const std::string& stat_prefix,
+                   std::unordered_set<RouteConfigProvider*>& route_config_providers,
+                   SubscriptionFactoryFunction factory_function =
+                   Envoy::Config::SubscriptionFactory::subscriptionFromConfigSource);
+  ~VhdsSubscription() override { init_target_.ready(); }
+
+  // Config::SubscriptionCallbacks
+  // TODO(fredlas) deduplicate
+  void onConfigUpdate(const Protobuf::RepeatedPtrField<ProtobufWkt::Any>&,
+                      const std::string&) override {
+    NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+  }
+  void onConfigUpdate(const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>&,
+                      const Protobuf::RepeatedPtrField<std::string>&, const std::string&) override;
+  void onConfigUpdateFailed(const EnvoyException* e) override;
+  std::string resourceName(const ProtobufWkt::Any& resource) override {
+    return MessageUtil::anyConvert<envoy::api::v2::route::VirtualHost>(resource).name();
+  }
+
+  void registerInitTargetWithInitManager(Init::Manager& m) { m.add(init_target_); }
+  void initializeVhosts(const envoy::api::v2::RouteConfiguration& route_configuration);
+  void removeVhosts(std::unordered_map<std::string, envoy::api::v2::route::VirtualHost>& vhosts,
+                    const Protobuf::RepeatedPtrField<std::string>& removed_vhost_names);
+  void updateVhosts(std::unordered_map<std::string, envoy::api::v2::route::VirtualHost>& vhosts,
+                    const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources);
+  void rebuildRouteConfig(
+          const std::unordered_map<std::string, envoy::api::v2::route::VirtualHost>& vhosts,
+          envoy::api::v2::RouteConfiguration& route_config);
+  absl::optional<LastConfigInfo> configInfo() const override { return config_info_; }
+  envoy::api::v2::RouteConfiguration& routeConfiguration() override { return route_config_proto_; }
+  SystemTime lastUpdated() const override { return last_updated_; }
+
+  std::unique_ptr<Envoy::Config::Subscription> subscription_;
+  envoy::api::v2::RouteConfiguration route_config_proto_;
+  const std::string route_config_name_;
+  Init::TargetImpl init_target_;
+  Stats::ScopePtr scope_;
+  VhdsStats stats_;
+  TimeSource& time_source_;
+  SystemTime last_updated_;
```

Comment:
It looks like last_updated_ and lastUpdated() aren't used for anything. Did I miss a use, or could it be removed?

Reply:
These are used here: https://github.com/envoyproxy/envoy/pull/6438/files#diff-a112a0982616fac9024fc1c8e42f0661R121. Please also see the thread at https://github.com/envoyproxy/envoy/pull/6438#discussion_r274124191.